Some combinations of intersections of circular kernel objest fail to compile

The following combinations fail: Line_2, Line_2, Line_arc_2, Circular_arc_2, and Circular_arc_2, Line_arc_2.
The attached source can be used to reproduce the problems.

[intersecting_arcs.cpp.zip](https://github.com/CGAL/cgal/files/611012/intersecting_arcs.cpp.zip)

I tried to modify a bit the code to plainly satisfy to the [reference manual](file:///home/simon/CGAL/master/build/release/doc_output/Kernel_23/group__intersection__circular__grp.html) and to be as explicit as possible (no automatic template deduction):

```c++
  typedef typename CGAL::cpp11::result_of<Circular_k::Intersect_2(Line_2, Line_2)>::type
    Intersection_result;
  std::vector<Intersection_result> res;
  CGAL::intersection<Line_2, Line_2, std::back_insert_iterator<std::vector<Intersection_result> > >(line, line, std::back_inserter(res));
```

It still doesn't compile but the error is different. It seems that the right version of `CGAL::intersection()` is never found, I have this kind of error:

```
note:   template argument deduction/substitution failed:
/home/simon/CGAL/My_tests/intersecting_arcs.cpp:25:136: error: wrong number of template arguments (3, should be 1)
   CGAL::intersection<Line_2, Line_2, std::back_insert_iterator<std::vector<Intersection_result> > >(line, line, std::back_inserter(res));
```

Even if I add explicitly `#include <CGAL/Circular_kernel_intersections.h>`, same error...

--- a/Circular_kernel_2/include/CGAL/Circular_kernel_2/Intersection_traits.h
+++ b/Circular_kernel_2/include/CGAL/Circular_kernel_2/Intersection_traits.h
@@ -153,6 +153,12 @@ struct CK2_Intersection_traits<CK, typename CK::Circle_2, typename CK::Line_2> :
     public CK2_Intersection_traits<CK, typename CK::Line_2, typename CK::Circle_2>
 {};
 
+template<typename CK>
+struct CK2_Intersection_traits<CK, typename CK::Line_2, typename CK::Line_2>
+{
+  typedef typename Intersection_traits<CK, typename CK::Line_2, typename CK::Line_2>::result_type type;
+};
+
 } //end of namespace CGAL
 
 #else

--- a/Circular_kernel_2/include/CGAL/Circular_kernel_2/function_objects_polynomial_circular.h
+++ b/Circular_kernel_2/include/CGAL/Circular_kernel_2/function_objects_polynomial_circular.h
@@ -632,6 +632,15 @@ namespace CircularFunctors {
 	       OutputIterator res) const
     { return CircularFunctors::intersect_2<CK> (l,la,res); }
 
+    template < class OutputIterator >
+    OutputIterator
+    operator()(const Line & l1, const Line & l2,
+	       OutputIterator res) const
+    {
+      *res++=typename CK::Linear_kernel::Intersect_2()(l1, l2);
+      return res;
+    }
+
   };
 
   template < class CK >

--- a/Circular_kernel_2/include/CGAL/Circular_kernel_2/internal_functions_on_line_arc_2.h
+++ b/Circular_kernel_2/include/CGAL/Circular_kernel_2/internal_functions_on_line_arc_2.h
@@ -557,9 +557,6 @@ namespace CircularFunctors {
 	       const typename CK::Circular_arc_2 &c,
 	       OutputIterator res )
   {
-    typedef std::vector<CGAL::Object > solutions_container;
-    typedef typename CK::Circular_arc_point_2 Circular_arc_point_2;
-
 #if defined(CGAL_CK_EXPLOIT_IDENTITY) || \
   defined(CGAL_INTERSECTION_MAP_FOR_SUPPORTING_CIRCLES)
     typedef typename CK::Line_arc_2 Line_arc_2;
@@ -667,7 +664,8 @@ namespace CircularFunctors {
     }
 #endif // CGAL_CK_EXPLOIT_IDENTITY
 
-    typedef std::vector<CGAL::Object > solutions_container;
+    typedef std::vector<typename CK2_Intersection_traits<CK, typename CK::Line_2, typename CK::Circle_2>::type>
+      solutions_container;
     solutions_container solutions;
 
 #ifdef CGAL_INTERSECTION_MAP_FOR_SUPPORTING_CIRCLES
@@ -687,25 +685,30 @@ namespace CircularFunctors {
       }
 #endif
 
+
     for (typename solutions_container::iterator it = solutions.begin();
-	 it != solutions.end(); ++it) {
-      const std::pair<Circular_arc_point_2, unsigned>
-        *result = CGAL::object_cast
-	  <std::pair<Circular_arc_point_2, unsigned> > (&(*it));
-#ifdef CGAL_CK_TEST_BBOX_BEFORE_HAS_ON
-	  Bbox_2 rb = result->first.bbox();
-	  if(do_overlap(l.bbox(), rb) && do_overlap(c.bbox(),rb)){
-	    if (has_on<CK>(l,result->first,true) &&
-		has_on<CK>(c,result->first,true)) {
-	      *res++ = *it;
-	    }
-	  }
-#else
-      if (has_on<CK>(l,result->first,true) &&
-          has_on<CK>(c,result->first,true)) {
+	 it != solutions.end(); ++it)
+    {
+      #if CGAL_INTERSECTION_VERSION  < 2
+      if(const std::pair<typename CK::Circular_arc_point_2, unsigned>* p =
+         object_cast< std::pair< typename CK::Circular_arc_point_2, unsigned> >(& (*it)))
+      {
+        #ifdef CGAL_CK_TEST_BBOX_BEFORE_HAS_ON
+        Bbox_2 rb = p->first.bbox();
+        if(!do_overlap(l.bbox(), rb) || !do_overlap(c.bbox(),rb)) continue;
+        #endif
+        Has_on_visitor<CK, typename CK::Line_arc_2> vis1(&l);
+        Has_on_visitor<CK, typename CK::Circular_arc_2> vis2(&c);
+        if(vis1(*p) && vis2(*p))
+          *res++ = *it;
+      }
+      #else
+      if(boost::apply_visitor(Has_on_visitor<CK, typename CK::Line_arc_2>(&l), *it) &&
+         boost::apply_visitor(Has_on_visitor<CK, typename CK::Circular_arc_2>(&c), *it) )
+      {
 	*res++ = *it;
       }
-#endif
+      #endif
     }
     return res;
   }

--- a/Circular_kernel_2/include/CGAL/Circular_kernel_intersections.h
+++ b/Circular_kernel_2/include/CGAL/Circular_kernel_intersections.h
@@ -86,6 +86,13 @@ intersection(const Line_2 <K> &c1, const Circle_2 <K> &c2, OutputIterator res)
   return typename K::Intersect_2()(c1, c2, res);
 }
 
+template < class OutputIterator, class K >
+OutputIterator
+intersection(const Line_2 <K> &c1, const Line_2 <K> &c2, OutputIterator res)
+{
+  return typename K::Intersect_2()(c1, c2, res);
+}
+
 CGAL_CIRCULAR_KERNEL_MACRO_GLOBAL_FUNCTION_INTERSECTION_(Circular_arc_2, Circular_arc_2)
 CGAL_CIRCULAR_KERNEL_MACRO_GLOBAL_FUNCTION_INTERSECTION_(Line_arc_2, Line_arc_2)
 CGAL_CIRCULAR_KERNEL_MACRO_GLOBAL_FUNCTION_INTERSECTION_(Line_arc_2, Circle_2)

